_overviews/core/string-interpolation.md

```diff
@@ -105,42 +127,106 @@ In Scala, all processed string literals are simple code transformations.   Anyti
+Here's an example:
 
-    // Note: We extends AnyVal to prevent runtime instantiation.  See
-    // value class guide for more info.
-    implicit class JsonHelper(val sc: StringContext) extends AnyVal {
-      def json(args: Any*): JSONObject = sys.error("TODO - IMPLEMENT")
-    }
+{% tabs json-definition-and-usage class=tabs-scala-version %}
+
+{% tab 'Scala 2' for=json-definition-and-usage %}
+```scala
+// Note: We extends AnyVal to prevent runtime instantiation.  See
+// value class guide for more info.
+implicit class JsonHelper(val sc: StringContext) extends AnyVal {
+  def json(args: Any*): JSONObject = sys.error("TODO - IMPLEMENT")
+}
+
+def giveMeSomeJson(x: JSONObject): Unit = ...
+
+giveMeSomeJson(json"{ name: $name, id: $id }")
+```
+{% endtab %}
+
+{% tab 'Scala 3' for=json-definition-and-usage %}
+```scala
+extension (sc: StringContext)
+  def json(args: Any*): JSONObject = sys.error("TODO - IMPLEMENT")
 
-    def giveMeSomeJson(x: JSONObject): Unit = ...
+def giveMeSomeJson(x: JSONObject): Unit = ...
 
-    giveMeSomeJson(json"{ name: $name, id: $id }")
+giveMeSomeJson(json"{ name: $name, id: $id }")
+```
+{% endtab %}
+
+{% endtabs %}
 
 In this example, we're attempting to create a JSON literal syntax using string interpolation.   The `JsonHelper` implicit class must be in scope to use this syntax, and the json method would need a complete implementation.   However, the result of such a formatted string literal would not be a string, but a `JSONObject`.
 
 When the compiler encounters the literal `json"{ name: $name, id: $id }"` it rewrites it to the following expression:
 
-    new StringContext("{ name: ", ", id: ", " }").json(name, id)
+{% tabs extension-desugaring class=tabs-scala-version %}
+
+{% tab 'Scala 2' for=extension-desugaring %}
+```scala
+new StringContext("{ name: ", ", id: ", " }").json(name, id)
+```
 
 The implicit class is then used to rewrite it to the following:
 
-    new JsonHelper(new StringContext("{ name: ", ", id: ", " }")).json(name, id)
-
-So, the `json` method has access to the raw pieces of strings and each expression as a value.   A simple (buggy) implementation of this method could be:
-
-    implicit class JsonHelper(val sc: StringContext) extends AnyVal {
-      def json(args: Any*): JSONObject = {
-        val strings = sc.parts.iterator
-        val expressions = args.iterator
-        var buf = new StringBuilder(strings.next())
-        while(strings.hasNext) {
-          buf.append(expressions.next())
-          buf.append(strings.next())
-        }
-        parseJson(buf)
-      }
+```scala
+new JsonHelper(new StringContext("{ name: ", ", id: ", " }")).json(name, id)
+```
+{% endtab %}
+
+{% tab 'Scala 3' for=extension-desugaring %}
+```scala
+new StringContext("{ name: ", ", id: ", " }").json(name, id)
+```
+
+The extension method is then used to rewrite it to the following:
+
+```scala
+json(new StringContext("{ name: ", ", id: ", " }"))(name, id)
+```
```

Comment:
Why do we want to add that?

Reply:
Because it corresponds to the explanation in Scala 2. But maybe it's more confusing than useful?

Reply:
Yes, I would leave this part because users don’t need to know that in this article. They only need to understand that their `json` method will be called when they write `json"something"`.